### PR TITLE
Extract computed keys from the class closure

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/8300
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The other two tests are bugs which have never been reported, but which are also fixed by this new computed keys approach.

cc @benjamn 

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13600"></a>

### CODE REVIEW

File: packages/babel-plugin-transform-classes/test/fixtures/spec/computed-keys-call-order/exec.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+var log = [];
+
+class A {
+  [log.push(1)]() {}
+  static [log.push(2)]() {}
+  [log.push(3)]() {}
+  static [log.push(4)]() {}
+}
+
+expect(log).toEqual([1, 2, 3, 4]);
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-classes/test/fixtures/spec/computed-keys-call-order/exec.js:10]: On `main` this is `1, 3, 2, 4`.

**Review by benjamn — APPROVED** [review]
LGTM (this should resolve https://github.com/facebook/regenerator/pull/487). Thanks @nicolo-ribaudo!

File: packages/babel-plugin-transform-classes/src/transformClass.ts [hunk 2]
```diff
@@ -600,8 +602,26 @@ export default function transformClass(
...
 
+  function extractDynamicKeys() {
+    const { dynamicKeys, node, scope } = classState as {
+      dynamicKeys: Map<string, t.Expression>;
+      node: t.Class;
+      scope: NodePath["scope"];
+    };
+
+    node.body.body.forEach(elem => {
+      if (!t.isClassMethod(elem) || !elem.computed) return;
```
Comment (JLHwung) [inline packages/babel-plugin-transform-classes/src/transformClass.ts:613]: Q: Why should we exclude non-method elements? I think we are preserving the order of `ClassElementName` evaluation and thus we should apply it to any named class elements.

**Review by JLHwung — APPROVED** [review]

File: packages/babel-plugin-transform-classes/src/transformClass.ts [hunk 3]
```diff
@@ -600,8 +602,26 @@ export default function transformClass(
...
   }
 
+  function extractDynamicKeys() {
+    const { dynamicKeys, node, scope } = classState as {
+      dynamicKeys: Map<string, t.Expression>;
+      node: t.Class;
+      scope: NodePath["scope"];
+    };
+
+    node.body.body.forEach(elem => {
```
Comment (KFlash) [inline packages/babel-plugin-transform-classes/src/transformClass.ts:612]: Foreach performance is approximately 6 times slower than FOR / FOREACH performance. The FOR loop without length caching works 3 times slower on lists, comparing to arrays.

For those who are interested in performance - ForEach is 96% slower than for loop..

https://stackoverflow.com/questions/43821759/why-array-foreach-is-slower-than-for-loop-in-javascript

File: packages/babel-plugin-transform-classes/src/transformClass.ts (hunk 2 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-classes/src/transformClass.ts:613]: Yeah, this plugin doesn't support class fields.

**Review by existentialism — APPROVED** [review]

**Pull request merged**